pkg/router/controller/status.go

```diff
@@ -0,0 +1,239 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	lru "github.com/hashicorp/golang-lru"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/sets"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/openshift/origin/pkg/client"
+	routeapi "github.com/openshift/origin/pkg/route/api"
+	"github.com/openshift/origin/pkg/router"
+)
+
+// StatusAdmitter ensures routes added to the plugin have status set.
+type StatusAdmitter struct {
+	plugin     router.Plugin
+	client     client.RoutesNamespacer
+	routerName string
+
+	contentionInterval time.Duration
+	expected           *lru.Cache
+}
+
+// NewStatusAdmitter creates a plugin wrapper that ensures every accepted
+// route has a status field set that matches this router. The admitter manages
+// an LRU of recently seen conflicting updates to handle when two router processes
+// with differing configurations are writing updates at the same time.
+func NewStatusAdmitter(plugin router.Plugin, client client.RoutesNamespacer, name string) *StatusAdmitter {
+	expected, _ := lru.New(1024)
+	return &StatusAdmitter{
+		plugin:     plugin,
+		client:     client,
+		routerName: name,
+
+		contentionInterval: 1 * time.Minute,
+		expected:           expected,
+	}
+}
+
+// nowFn allows the package to be tested
+var nowFn = unversioned.Now
+
+// findOrCreateIngress loops through the router status ingress array looking for an entry
+// that matches name. If there is no entry in the array, it creates one and appends it
+// to the array. If there are multiple entries with that name, the first one is
+// returned and later ones are removed. Changed is returned as true if any part of the
+// array is altered, and hostChanged is true if the host value was changed.
+func findOrCreateIngress(route *routeapi.Route, name string) (_ *routeapi.RouteIngress, changed bool) {
+	position := -1
+	updated := make([]routeapi.RouteIngress, 0, len(route.Status.Ingress))
+	for i := range route.Status.Ingress {
+		existing := &route.Status.Ingress[i]
+		if existing.RouterName != name {
+			updated = append(updated, *existing)
+			continue
+		}
+		if position != -1 {
+			changed = true
+			continue
+		}
+		position = i
+	}
+	switch {
+	case position == -1:
+		position = len(route.Status.Ingress)
+		route.Status.Ingress = append(route.Status.Ingress, routeapi.RouteIngress{
+			RouterName: name,
+			Host:       route.Spec.Host,
+		})
+		changed = true
+	case changed:
+		route.Status.Ingress = updated
```

Comment:
Wouldn't this assignment cause problems - because if `position` is set to a non -1 value and changed is true, the array index on line 82 below would cause problems because `route.Status.Ingress` is updated [no pun intended] here.